python/src/equistore/operations/_utils.py

```diff
@@ -48,58 +48,121 @@ def _check_blocks(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
+                    raise ValueError(err_msg + err_msg_1)
+
         else:
             raise ValueError(
                 f"{prop} is not a valid property to check. "
-                "Choose from 'samples', 'properties', 'components' "
-                "or 'gradients'."
+                "Choose from 'samples', 'properties', 'components'."
             )
 
 
-def _check_same_gradients_components(a: TensorBlock, b: TensorBlock, fname: str):
-    for parameter, grad_a in a.gradients():
-        grad_b = b.gradient(parameter)
-        grad_comps_a = np.array(grad_a.components, dtype=object)
-        grad_comps_b = np.array(grad_b.components, dtype=object)
+def _check_same_gradients(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
+    """Check if metadata between two gradients's TensorBlocks is consistent
+     for an operation.
 
-        if not np.all(grad_comps_a == grad_comps_b):
-            raise ValueError(
-                f"input to {fname} should habe the same gradient components"
-            )
+    The functions verifies that that the metadata of the given props is the same
+    (length, names, values).
+
+    If props is None it only checks if the ``'parameters'`` are consistent.
+
+    :param a: first :py:class:`TensorBlock` for check
+    :param b: second :py:class:`TensorBlock` for check
+    :param props: A list of strings containing the property to check.
+                 Allowed values are ``'samples'`` or ``'properties'``,
+                 ``'components'``. To check only if the ``'parameters'`` are consistent
+                 use ``props=None``, using ``'parameters'`` is allowed
+                  even though deprecated.
+    """
+    err_msg = f"Inputs to {fname} should have the same gradient:\n"
+    grads_a = a.gradients_list()
+    grads_b = b.gradients_list()
+
+    if len(grads_a) != len(grads_b) or (
+        not np.all([parameter in grads_b for parameter in grads_a])
+    ):
+        raise ValueError(f"Inputs to {fname} should have the same gradient parameters.")
+
+    if props is not None:
+        for parameter, grad_a in a.gradients():
+            grad_b = b.gradient(parameter)
+            for prop in props:
+                err_msg_len = (
+                    f'gradient ("{parameter}") {prop} of the two `TensorBlock` '
+                    "have not the same lenght."
+                )
+
+                err_msg_1 = (
+                    f'gradient ("{parameter}") {prop} are not the same or not in the '
+                    "same order."
+                )
+
+                err_msg_names = (
+                    f'gradient ("{parameter}") {prop} names are not the same '
+                    "or not in the same order."
+                )
+
+                if prop == "samples":
+                    if not len(grad_a.samples) == len(grad_b.samples):
+                        raise ValueError(err_msg + err_msg_len)
+                    if not grad_a.samples.names == grad_b.samples.names:
```

Comment:
Do I see this correct that we check for exactly the same gradients. If the values are shuffled in one gradients wrt to the second we throw an error. For me this is fine, but we should state this in the docstring.

Reply:
Now these functions always check for the values and the order. All the operations that uses this check so far are implemented that need the order. Make them work non with the order is the next step

Reply:
i added a comment in the docstring